source/common/quic/envoy_quic_server_stream.cc

```diff
@@ -257,19 +257,32 @@ void EnvoyQuicServerStream::maybeDecodeTrailers() {
 }
 
 bool EnvoyQuicServerStream::OnStopSending(quic::QuicRstStreamErrorCode error) {
+  // Only called in IETF Quic to close write side.
+  ENVOY_STREAM_LOG(debug, "received STOP_SENDING with reset code={}", *this, error);
+  stats_.rx_reset_.inc();
+  bool end_stream_encoded = local_end_stream_;
+  // This call will close write.
   bool ret = quic::QuicSpdyServerStreamBase::OnStopSending(error);
-  if (read_side_closed()) {
+  if (read_side_closed() && !end_stream_encoded) {
+    // If both directions are closed but end stream hasn't be encoded yet, notify reset callbacks.
     // Treat this as a remote reset, since the stream will be closed in both directions.
     runResetCallbacks(quicRstErrorToEnvoyRemoteResetReason(error));
   }
   return ret;
 }
 
 void EnvoyQuicServerStream::OnStreamReset(const quic::QuicRstStreamFrame& frame) {
-  ENVOY_STREAM_LOG(debug, "received reset code={}", *this, frame.error_code);
+  ENVOY_STREAM_LOG(debug, "received RESET_STREAM with reset code={}", *this, frame.error_code);
   stats_.rx_reset_.inc();
+  bool end_stream_decoded_and_encoded = read_side_closed() && local_end_stream_;
```

Comment:
I think that the value of local_end_stream_ is ignored in this method.

This is easiest to deduce in the IETF QUIC case.  I assume that it is not possible for read_side_closed and write_side_closed to both be true when you enter this method.  If write_side_closed is not expected to change when OnStreamReset is called for IETF QUIC, then this method is only entered if read_side_closed() would return false.  If read_side_closed() is false, the value of local_end_stream_ doesn't matter.

In the case of gQUIC, I guess local_end_stream_ would matter in the case this method is entered when read_side_closed is true.  If read_side_closed is false, I think you'ld end up in the situation where reset callbacks are invoked even after local_end_stream_ is set which may cause problems.

It would be good to see if it is possible to enter this method when write_side_closed is true, read_side_closed is false and local_end_stream_ is true.

Reply:
> I think that the value of local_end_stream_ is ignored in this method.
> 
> This is easiest to deduce in the IETF QUIC case. I assume that it is not possible for read_side_closed and write_side_closed to both be true when you enter this method. If write_side_closed is not expected to change when OnStreamReset is called for IETF QUIC, then this method is only entered if read_side_closed() would return false. If read_side_closed() is false, the value of local_end_stream_ doesn't matter.

This method can be entered when both write_side_closed and read_side_closed are true. In the newly added integration test,  STOP_SENDING and RESET_STREAM are received sequentially. The read side is already closed by then and STOP_SENDING closes write side. This method is still triggered in that case.

> In the case of gQUIC, I guess local_end_stream_ would matter in the case this method is entered when read_side_closed is true. If read_side_closed is false, I think you'ld end up in the situation where reset callbacks are invoked even after local_end_stream_ is set which may cause problems.
> 
> It would be good to see if it is possible to enter this method when write_side_closed is true, read_side_closed is false and local_end_stream_ is true.

Yes, in testRouterUpstreamResponseBeforeRequestComplete() with Http3 upstream, a quic server stream will encode end_stream and close write side before read side gets closed by Reset. In that case, we want to invoke reset callbacks upon receiving that Reset even local_end_stream_ is set. 

In general, for both  OnStreamReset() and OnStopSending() we want to avoid calling reset callbacks after we both encoded and decoded end_stream.

Reply:
I think I'll have to keep this code in mind if the HCM implementation is changed to not reset the stream in the case where read_closed is false and local_end_stream is true due to some early reply cases.  I guess this is fine for now and is likely to improve once gQUIC is removed.